src/main/java/org/alliancegenome/curation_api/services/GeneExpressionExperimentService.java

```diff
@@ -0,0 +1,72 @@
+package org.alliancegenome.curation_api.services;
+
+import jakarta.annotation.PostConstruct;
+import jakarta.enterprise.context.RequestScoped;
+import jakarta.inject.Inject;
+import jakarta.transaction.Transactional;
+import lombok.extern.jbosslog.JBossLog;
+import org.alliancegenome.curation_api.dao.GeneExpressionAnnotationDAO;
+import org.alliancegenome.curation_api.dao.GeneExpressionExperimentDAO;
+import org.alliancegenome.curation_api.exceptions.ValidationException;
+import org.alliancegenome.curation_api.model.entities.GeneExpressionAnnotation;
+import org.alliancegenome.curation_api.model.entities.GeneExpressionExperiment;
+import org.alliancegenome.curation_api.response.SearchResponse;
+import org.alliancegenome.curation_api.services.base.BaseEntityCrudService;
+import org.alliancegenome.curation_api.services.ontology.MmoTermService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+@JBossLog
+@RequestScoped
+public class GeneExpressionExperimentService extends BaseEntityCrudService<GeneExpressionExperiment, GeneExpressionExperimentDAO> {
+
+	@Inject GeneExpressionExperimentDAO geneExpressionExperimentDAO;
+	@Inject GeneExpressionAnnotationDAO geneExpressionAnnotationDAO;
+	@Inject GeneService geneService;
+	@Inject MmoTermService mmoTermService;
+	@Inject ReferenceService referenceService;
+
+	@Override
+	@PostConstruct
+	protected void init() {
+		setSQLDao(geneExpressionExperimentDAO);
+	}
+
+	@Transactional
+	public GeneExpressionExperiment upsert(String experimentId, Set<String> geneExpressionAnnotationIds) throws ValidationException {
+		GeneExpressionExperiment geneExpressionExperiment;
+		List<GeneExpressionAnnotation> annotations;
+
+		//	example of experimentId: Xenbase:XB-GENE-972235|AGRKB:101000000874667|MMO:0000658
+		String[] definingFields = experimentId.split("\\|", 3);
+		String geneId = definingFields[0];
+		String referenceId = definingFields[1];
+		String assayId = definingFields[2];
+		SearchResponse<GeneExpressionExperiment> response = geneExpressionExperimentDAO.findByField("uniqueId", experimentId);
+		if (response != null && response.getSingleResult() != null) {
+			geneExpressionExperiment = response.getSingleResult();
+		} else {
+			geneExpressionExperiment = new GeneExpressionExperiment();
+			geneExpressionExperiment.setUniqueId(experimentId);
+		}
+
+		geneExpressionExperiment.setEntityAssayed(geneService.findByIdentifierString(geneId));
```

Comment:
I don't see any validation of these fields.  Is this relying on the components being validated prior to generating the experimentId?  If so, could that be fragile and are there alternative routes into this method such as an API request?

Reply:
All fields were validated when loading the expression annotations. 
About fragility: if the load process is interrupted, the experiments cached in memory are lost. The only way to fix this would be a reload.

Reply:
What happens if the annotations failed validation - does the code take this into account and not attempt to create experiments for failed annotations?

Reply:
If the annotation was not valid, the experiment (paperId, assayId, geneId) is not created